core/deployment/src/main/java/io/quarkus/deployment/ExtensionLoader.java

```diff
@@ -214,7 +218,45 @@ public boolean canHandleObject(final Object obj, final boolean staticInit) {
                 .produces(RunTimeConfigurationProxyBuildItem.class)
                 .produces(BytecodeRecorderObjectLoaderBuildItem.class)
                 .build());
-        return result;
+
+        return result.andThen(provideCapabilities(classLoader, appModel, readResult, launchMode, devModeType));
+    }
+
+    private static Consumer<BuildChainBuilder> provideCapabilities(ClassLoader classLoader, AppModel appModel,
```

Comment:
Could this loop not be done as a build step which produces a `CapabilityBuildItem`?  This kind of change to the extension loader makes it more brittle, and it was definitely intended that the build item system be able to support this kind of thing.

Reply:
I changed the loop to add the build steps to the `result` directly but I am not sure whether that's what you meant. Could you clarify please? Thanks.

Reply:
The `provideCapabilities` method seems to produce one `CapabilityBuildItem` for each discovered capability.  Why not do this in a build step, which are able to examine the build class path and produce build items?

Reply:
Could you please review the last commit? Also I noticed the condition cache (`condCache`) is created per `buildIf` call. Should we create a global one, e.g. in `public static Consumer<BuildChainBuilder> loadStepsFrom(ClassLoader classLoader, Properties buildSystemProps,
            AppModel appModel, LaunchMode launchMode, DevModeType devModeType,
            Consumer<ConfigBuilder> configCustomizer)`? Thanks!